### PR TITLE
DEL-3317: Add the runtime descriptor generation mechanism on the distro assembly

### PR DESCRIPTION
**Pull request by garodriguezlp**
- Add the assembly descriptor generation Mojo

### CODE REVIEW

File: src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyRepackager.groovy [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+        if (!targetDir.exists() && !targetDir.mkdirs()) {
+            throw new IOException("Can't create dir ${targetDir} to place the descriptor file")
+        }
+
+        File descriptorArchiveNewHome = new File(targetDir, descriptorArchive.name)
+        descriptorArchiveNewHome << descriptorArchive.bytes
+    }
+
+    private repackageAssembly(File assemblyFile, File assemblyRootDir) {
```
Comment (aiannucci) [inline src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyRepackager.groovy:71]: Should we move this out of the plugin in order to properly separate concerns?
The plugin could only generate the assembly descriptor, and then we repackage the distribution to add it by using another maven assembly definition.

File: src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorValidator.groovy [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    }
+
+    static void validateDescriptorTempDir(File descriptorTempDir) {
+        if (descriptorTempDir.exists() && !descriptorTempDir.isDirectory()) {
+            throw new IllegalArgumentException("Given descriptor generation temp dir path ${descriptorTempDir} is not a " +
+                    "directory")
+        }
+
+        if (!descriptorTempDir.exists() && !descriptorTempDir.mkdirs()) {
+            throw new IllegalArgumentException("Can not create descriptor generation temp dir at ${descriptorTempDir}")
```
Comment (aiannucci) [inline src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorValidator.groovy:18]:
```suggestion
            throw new IllegalArgumentException("Cannot create descriptor generation temp dir at ${descriptorTempDir}")
```

File: src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    private void printSplash() {
+        log.info "*" * 80
+        log.info("Generating descriptor for the assembly".center(80))
+        log.info "*" * 80
+    }
+
+    private File setupWorkingDirForExecution() {
+        // @todo[question]: Should this explanation be left?
```
Comment (aiannucci) [inline src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy:94]: No, i don't see it as a hack anyways, it's called proper coding

File: src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            File contentDescriptor = new AssemblyContentDescriptorGenerator(log: log, workingDir: descriptorTempDir)
+                    .generateDescriptor(assemblyFile)
+
+            // @todo[question]: Would clients "scanners" be scared by such a new particular jar? Wouldn't be less scared if
+            //  seeing a plain file?
+            File descriptorArchive = new AssemblyDescriptorArchiveBuilder(log: log, workingDir: descriptorTempDir)
+                    .buildDescriptorArchive(contentDescriptor, descriptorJarName)
+
+            // @todo[question]: How inconvenient is the idea of repackaging the distro? Let's say from the Maven's convention
+            //  perspective, How many conventions might we braking with this?
```
Comment (aiannucci) [inline src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy:77]: I would leave the responsibility of repackaging out of this plugin, and introduce that in a separate maven assembly descriptor, to create the final distribution artifact that includes the descriptor.

File: src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        try {
+            AssemblyDescriptorValidator.validateAssemblyFile(assemblyFile)
+            AssemblyDescriptorValidator.validateDescriptorTempDir(descriptorTempDir)
+            descriptorTempDir = setupWorkingDirForExecution()
+
+            File contentDescriptor = new AssemblyContentDescriptorGenerator(log: log, workingDir: descriptorTempDir)
+                    .generateDescriptor(assemblyFile)
+
+            // @todo[question]: Would clients "scanners" be scared by such a new particular jar? Wouldn't be less scared if
+            //  seeing a plain file?
```
Comment (aiannucci) [inline src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy:72]: Why would they be scared of a JAR?

File: src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy [hunk 6]
```diff
@@ -0,0 +1,103 @@
+// @todo[question]: Should we add any copyright headers
```
Comment (aiannucci) [inline src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy:1]: Yes

File: src/main/groovy/org/mule/tools/assembly/compress/ArchiveUtils.groovy [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        }
+        return entries
+    }
+
+    private static List listArchiveEntries(ArchiveInputStream stream, File archive) {
+        List entries = []
+        ArchiveEntry entry
+        while ((entry = stream.getNextEntry()) != null) {
+            if (!stream.canReadEntryData(entry)) {
+                throw new IllegalStateException("Cant ready entry ${entry.name} from ${archive}")
```
Comment (aiannucci) [inline src/main/groovy/org/mule/tools/assembly/compress/ArchiveUtils.groovy:115]:
```suggestion
                throw new IllegalStateException("Cannot read entry ${entry.name} from ${archive}")
```

File: src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+     * Version of the product used for the descriptor zip file name
+     */
+    @Parameter(defaultValue = 'mule-assembly-descriptor-${project.version}.jar')
+    String descriptorJarName
+
+    /**
+     * Relative path of the location where the descriptor jar should be placed within the assembly's root directory
+     */
+    @Parameter(defaultValue = 'lib/mule')
+    String targetDescriptorDirWithinAssembly
```
Comment (aiannucci) [inline src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy:46]: This shouldn't be necessary if we move the repackaging logic out

**Review by aiannucci — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 9]
```diff
@@ -30,18 +29,32 @@
         <gmavenplus-plugin.version>1.12.1</gmavenplus-plugin.version>
         <maven-invoker-plugin.version>3.2.2</maven-invoker-plugin.version>
         <jacoco-maven-plugin.version>0.8.7</jacoco-maven-plugin.version>
+        <slf4j-simple.version>1.7.31</slf4j-simple.version>
+        <commons-compress.version>1.20</commons-compress.version>
+        <jackson.version>2.12.4</jackson.version>
     </properties>
 
     <dependencies>
+        <!-- // @todo[question]: What should we do with https://github.com/mulesoft/mule/blob/fae2e775239109866c6921d627fa6ac18e6e865a/pom.xml#L1581 ?-->
```
[the remark this answers is not in the document]
Comment (garodriguezlp) [reply pom.xml:38]: Great! I'll send the PR with the adjustment once we're done with the changes on this side and have released the new version

File: src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy (hunk 5 above)
Comment (garodriguezlp) [reply src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy:77]: You're right. I think that was an unjustified panic attack 😅

File: src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy (hunk 8 above)
Comment (garodriguezlp) [reply src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy:51]: You're right!

File: src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy (hunk 6 above)
Comment (garodriguezlp) [reply src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy]: Done

File: src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy (hunk 4 above)
Comment (garodriguezlp) [reply src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy:82]: Got it! Let's do that

File: src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyRepackager.groovy (hunk 1 above)
Comment (garodriguezlp) [reply src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyRepackager.groovy]: Sure!

File: src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy (hunk 3 above)
Comment (garodriguezlp) [reply src/main/groovy/org/mule/tools/assembly/descriptor/AssemblyDescriptorGeneratorMojo.groovy]: Great! 😅.  but I'm afraid with the new approach this setup won't be necessary

**Review by aiannucci — APPROVED** [review]

**Pull request merged**